src/api-binder.ts

```diff
@@ -0,0 +1,890 @@
+					'Trying to start poll without initializing API client',
+				);
+			}
+			this.pollTargetState();
+			return null;
+		});
+	}
+
+	public startCurrentStateReport() {
+		if (this.balenaApi == null) {
+			throw new InternalInconsistencyError(
+				'Trying to start state reporting without initializing API client',
+			);
+		}
+		this.deviceState.on('change', () => {
+			if (!this.reportPending) {
+				// A latency of 100ms should be acceptable and
+				// allows avoiding catching docker at weird states
+				this.reportCurrentState();
+			}
+		});
+		return this.reportCurrentState();
+	}
+
+	private getStateDiff(): DeviceApplicationState {
+		const lastReportedLocal = this.lastReportedState.local;
+		const lastReportedDependent = this.lastReportedState.dependent;
+		if (lastReportedLocal == null || lastReportedDependent == null) {
+			throw new InternalInconsistencyError(
+				`No local or dependent component of lastReportedLocal in ApiBinder.getStateDiff: ${JSON.stringify(
+					this.lastReportedState,
+				)}`,
+			);
+		}
+
+		const diff = {
+			local: _(this.stateForReport.local)
+				.omitBy((val, key: keyof DeviceApplicationState['local']) =>
+					_.isEqual(lastReportedLocal[key], val),
+				)
+				.omit(INTERNAL_STATE_KEYS)
+				.value(),
+			dependent: _(this.stateForReport.dependent)
+				.omitBy((val, key: keyof DeviceApplicationState['dependent']) =>
+					_.isEqual(lastReportedDependent[key], val),
+				)
+				.omit(INTERNAL_STATE_KEYS)
+				.value(),
+		};
+
+		return _.omitBy(diff, _.isEmpty);
+	}
+
+	private async sendReportPatch(
+		stateDiff: DeviceApplicationState,
+		conf: { apiEndpoint: string; uuid: string },
+	) {
+		if (this.cachedBalenaApi == null) {
+			throw new InternalInconsistencyError(
+				'Attempt to send report patch without an API client',
+			);
+		}
+
+		const endpoint = url.resolve(
+			conf.apiEndpoint,
+			`/device/v2/${conf.uuid}/state`,
+		);
+
+		const requestParams = _.extend(
+			{
+				method: 'PATCH',
+				url: endpoint,
+				body: stateDiff,
+			},
+			this.cachedBalenaApi.passthrough,
+		);
+
+		await this.cachedBalenaApi._request(requestParams);
+	}
+
+	private async report() {
+		const conf = await this.config.getMany([
+			'deviceId',
+			'apiTimeout',
+			'apiEndpoint',
+			'uuid',
+			'localMode',
+		]);
+
+		if (checkTruthy(conf.localMode || false)) {
+			return;
+		}
+
+		const stateDiff = this.getStateDiff();
+		if (_.size(stateDiff) === 0) {
+			return 0;
+		}
+
+		await Bluebird.resolve(
+			this.sendReportPatch(stateDiff, conf as {
+				uuid: string;
+				apiEndpoint: string;
+			}),
+		).timeout(conf.apiTimeout as number);
+
+		this.stateReportErrors = 0;
+		_.assign(this.lastReportedState.local, stateDiff.local);
+		_.assign(this.lastReportedState.dependent, stateDiff.dependent);
+	}
+
+	private reportCurrentState(): null {
+		this.reportPending = true;
+		// TODO: Once device-state has been converted to typescript,
+		// remove this any
+		this.deviceState
```

Comment:
If you want you can probably keep this with await/async whilst still keeping the same semantics by using
```ts
(async () => {
	...
})()
return null
```